File: dsa-practice-ram/dsa-practice-ram/Arrays/TwoSum.cs
namespace dsa_practice_ram.Arrays
{
    internal class TwoSum
    {
        public void FindTwoSum()
        {
            int[] array = { 2, 6, 4, 6, 8, 5, 8, 1, 1, 10 };
            int target = 14;
            var pairs = new List<(int x, int y)>();
            var seen = new HashSet<int>();

            foreach (var x in array)
            {
                int y = target - x;

                if (seen.Contains(y))
                {
                    if (!pairs.Contains((y, x)) && !pairs.Contains((x, y)))
                        pairs.Add((x, y));
                }
                seen.Add(x);
            }

            foreach (var pair in pairs)
            {
                Console.WriteLine($"({pair.x}, {pair.y})");
            }
        }

        public void TwoSumOpimal()
        {
            int[] array = { 2, 6, 4, 6, 8, 5, 8, 1, 1, 10 };
            // sorted: { 1, 1, 2, 4, 5, 6, 8, 8, 10 }
            Array.Sort(array);

            int target = 14;

            int left = 0;
            int right = array.Length - 1;
            List<(int x, int y)> pairs = new List<(int x, int y)>();

            while (left < right)
            {
                int x = array[left];
                int y = array[right];

                if (x + y == target)
                {
                    pairs.Add((x, y));
                    left++;
                    right--;
                }
                else if (x + y < target)
                {
                    left++;
                }
                else
                    right--;
            }

            foreach (var pair in pairs)
            {
                Console.WriteLine($"({pair.x}, {pair.y})");
            }
        }
    }
}

File: MyLearning/PythonPractice/dsa_practice/arrays/two_sum.py
"""Two-sum pair finding algorithms."""

# C# TwoSum.FindTwoSum

def find_two_sum(array=None, target=14):
    """Return unique pairs that sum to target using a hash set.

    Problem: Find pairs of values that sum to a target.
    Approach: Track seen values and avoid duplicate pairs.
    Time complexity: O(n * p) for duplicate checks, typically O(n)
    Space complexity: O(n)
    """
    if array is None:
        array = [2, 6, 4, 6, 8, 5, 8, 1, 1, 10]

    pairs = []
    seen = set()

    for x in array:
        y = target - x
        if y in seen:
            if (y, x) not in pairs and (x, y) not in pairs:
                pairs.append((x, y))
        seen.add(x)

    return pairs


# C# TwoSum.TwoSumOpimal

def two_sum_opimal(array=None, target=14):
    """Return pairs that sum to target using sorting and two pointers.

    Problem: Find pairs of values that sum to a target.
    Approach: Sort the array and move two pointers inward.
    Time complexity: O(n log n)
    Space complexity: O(n) for the sorted copy
    """
    if array is None:
        array = [2, 6, 4, 6, 8, 5, 8, 1, 1, 10]

    nums = list(array)
    nums.sort()

    left = 0
    right = len(nums) - 1
    pairs = []

    while left < right:
        x = nums[left]
        y = nums[right]
        total = x + y

        if total == target:
            pairs.append((x, y))
            left += 1
            right -= 1
        elif total < target:
            left += 1
        else:
            right -= 1

    return pairs
